File: agents/empire-task-agent/agent/lib/tool-forge-mcp.ts
import { createEmpireMcpClient } from "#lib/mcp-client";

const HARVEST_CACHE_DIR =
  process.env.EMPIRE_HARVEST_CACHE_DIR ??
  "C:/Empire_Workbench/04_Thought_Experiments/harvest_cache";
const ACTIVE_TOOLS_DIR =
  process.env.EMPIRE_ACTIVE_TOOLS_DIR ?? "C:/Empire_Workbench/03_Active_Tools";

const mcp = createEmpireMcpClient({
  label: "empire-tool-forge",
  clientName: "eve-empire-tool-forge",
  script: "tool_forge_mcp.py",
  env: () => ({
    EMPIRE_HARVEST_CACHE_DIR: HARVEST_CACHE_DIR,
    EMPIRE_ACTIVE_TOOLS_DIR: ACTIVE_TOOLS_DIR,
  }),
});

export const connectEmpireToolForgeMcp = (): Promise<void> => mcp.connect();
export const disconnectEmpireToolForgeMcp = (): Promise<void> =>
  mcp.disconnect();

export async function docsGuideScrapeViaMcp(input: {
  root_url: string;
  max_pages?: number;
  note?: string;
}): Promise<unknown> {
  return mcp.callTool("docs_guide_scrape", {
    root_url: input.root_url,
    max_pages: input.max_pages ?? 80,
    note: input.note ?? "",
  });
}

export async function skillTriageManifestViaMcp(input: {
  paths?: string;
  include_installed?: boolean;
}): Promise<unknown> {
  return mcp.callTool("skill_triage_manifest", {
    paths: input.paths ?? "",
    include_installed: input.include_installed ?? true,
  });
}

export async function listHarvestOutputsViaMcp(): Promise<unknown> {
  return mcp.callTool("list_harvest_outputs", {});
}

File: mcp/tool_forge_mcp.py
"""FastMCP: Tool Forge — docs scrape, skill inventory, active tool listing."""

from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Any

from mcp.server.fastmcp import FastMCP

from pipeline import docs_guide_scraper, skill_compiler

mcp = FastMCP("empire-tool-forge")

ACTIVE_TOOLS_DIR = Path(
    os.environ.get("EMPIRE_ACTIVE_TOOLS_DIR", r"C:\Empire_Workbench\03_Active_Tools")
)
HARVEST_CACHE_DIR = Path(
    os.environ.get(
        "EMPIRE_HARVEST_CACHE_DIR",
        r"C:\Empire_Workbench\04_Thought_Experiments\harvest_cache",
    )
)


def _json(data: Any) -> str:
    return json.dumps(data, indent=2, default=str)


@mcp.tool()
async def docs_guide_discover(root_url: str, max_pages: int = 80) -> str:
    """Discover documentation page URLs via llms.txt or sitemap (no full scrape)."""
    return _json(docs_guide_scraper.discover_pages(root_url, max_pages=max_pages))


@mcp.tool()
async def docs_guide_scrape(root_url: str, max_pages: int = 80, note: str = "") -> str:
    """Scrape a documentation site into one Markdown guide under harvest_cache."""
    return _json(
        docs_guide_scraper.scrape_site(
            root_url,
            max_pages=max_pages,
            cache_dir=HARVEST_CACHE_DIR,
            note=note,
        )
    )


@mcp.tool()
async def skill_inventory(paths: str = "", include_installed: bool = True) -> str:
    """Inventory SKILL.md files from comma-separated paths/zips and .cursor/skills."""
    path_list = [p.strip() for p in paths.split(",") if p.strip()] if paths else []
    return _json(
        skill_compiler.inventory_sources(
            path_list,
            include_default_skills=include_installed,
        )
    )


@mcp.tool()
async def skill_triage_manifest(paths: str = "", include_installed: bool = True) -> str:
    """Run Build1 3-Bin heuristic triage; write manifest JSON + markdown under harvest_cache."""
    path_list = [p.strip() for p in paths.split(",") if p.strip()] if paths else []
    return _json(
        skill_compiler.run_inventory_and_triage(
            path_list,
            out_dir=HARVEST_CACHE_DIR,
            include_default_skills=include_installed,
        )
    )


@mcp.tool()
async def list_harvest_outputs() -> str:
    """List Markdown guides and triage manifests in harvest_cache."""
    HARVEST_CACHE_DIR.mkdir(parents=True, exist_ok=True)
    files = sorted(
        [
            {
                "name": p.name,
                "path": str(p),
                "bytes": p.stat().st_size,
            }
            for p in HARVEST_CACHE_DIR.iterdir()
            if p.is_file() and p.suffix.lower() in {".md", ".json"}
        ],
        key=lambda row: row["name"],
    )
    return _json({"ok": True, "dir": str(HARVEST_CACHE_DIR), "files": files, "count": len(files)})


@mcp.tool()
async def list_active_tools() -> str:
    """List flattened tool files in 03_Active_Tools (read-only index)."""
    if not ACTIVE_TOOLS_DIR.is_dir():
        return _json({"ok": False, "error": f"directory missing: {ACTIVE_TOOLS_DIR}"})
    files = sorted(
        p.name for p in ACTIVE_TOOLS_DIR.iterdir() if p.is_file() and not p.name.startswith(".")
    )
    return _json({"ok": True, "dir": str(ACTIVE_TOOLS_DIR), "files": files, "count": len(files)})


if __name__ == "__main__":
    mcp.run()
